### ISSUE TITLE
Builtin Converse.js: Better styling

### ISSUE THREAD

**Issue by drequivalent** [JohnXLivingston/peertube-plugin-livechat#72]
![Снимок экрана от 2021-11-05 17-26-09](https://user-images.githubusercontent.com/1463277/140527158-f91e3439-88bf-4d5d-9c1a-622a241f0f80.png)

1. Disable the avatars. They are always empty, they serve no purpose and their column steals the vertical screen space.
2. Also, that horizontal scroll on the message box has no reason to exist.
3. Also, the close button is too large, and the empty bar that it's on
 1) serves no purpose
 2) seems to reside ON TOP OF the chat log frame (which causes first messages to be visually cut off, and generally looks terrible)
Make it more subtle. Or get rid of it completely, actually.
4. Let the chat window take all the remaining space on the side of the video
5. Use full Unicode fonts.
---
6. Don't destroy the chat client when the window closes. Just hide it. Just so it doesn't have to connect and log in again.
7. Avoid converse.js splash screen

**Comment by JohnXLivingston** [comment]
Thanks for your contribution.

For now, I use the default ConverseJS templates and stylings.
I have planned to work on the UX/UI next week.

Some responses to your points:

1. Good point. Maybe some day I will enable external connection from XMPP clients, so there might be avatars. Or I could eventually use Peertube's avatar for connected users. I think that the best solution is to hide avatars depending on the chat width. When you have a little screen (on mobiles for example), the cat is underneath the video. So avatars are not a problem. Samewhen you open the chat in a new window. So hiding the avatar column depending on the chat window width seems to be a good solution.

2. I know. I just hadn't the time to correct this for now. Will be fixed next week.

3. The problem is that the buttons have to be outside the iframe. There is definitively some work to do to make this better. 

4. Same as 2.

5. Can you recommend me some standard fonts? Are the fonts used by Peertube correct? (for now, I'm using ConverseJS's fonts).

6. I have to think about it. It can have some drawbacks and/or bugs. Maybe adding a «minimize» button? To let the user choose to close or minimize?

7. That's on my TODO list. I already have a logo for the splash screen : https://github.com/JohnXLivingston/peertube-plugin-livechat/blob/main/public/images/logo.svg


Just a question about your screenshot. The button «open in a new window» is not here (it should be on the left of the close button). Have you disabled it? Or is it a bug?

**Comment by Horhik** [comment]
So, I made one shitty hotfix. It is using as a custom code to add to your peertube instance. 
Maybe this will help you somehow.


``` javascript
let first = true
setTimeout(() => {

let chatOpenButton = document.querySelector('.peertube-plugin-livechat-button-open')
let chatCloseButton = document.querySelector('.peertube-plugin-livechat-button-close')

let ccb = chatCloseButton.cloneNode(true);
chatCloseButton.parentNode.replaceChild(ccb, chatCloseButton);
// replacing "close chat button" to remove eventlisteners

// when "open chat button pressed"
document.querySelector('.peertube-plugin-livechat-button-open').addEventListener('click', (event) => {
    document.querySelector('#peertube-plugin-livechat-container iframe').style.display = 'block'

    document.querySelector('.peertube-plugin-livechat-button-open').style.display='none'
    document.querySelector('.peertube-plugin-livechat-button-close').style.display='block'


    document.querySelector('#video-wrapper').style.justifyContent = 'revert'
    document.querySelector('my-plugin-placeholder').style.flexGrow='1'
    document.querySelector('#videojs-wrapper').style.flexGrow='0'
})

// when "close chat button" pressed
document.querySelector('.peertube-plugin-livechat-button-close').addEventListener('click', (event) => {
  if(document.querySelector('#peertube-plugin-livechat-container iframe')){

    document.querySelector('#peertube-plugin-livechat-container iframe').style.display = 'none'
    document.querySelector('.peertube-plugin-livechat-button-open').style.display='block'
    document.querySelector('.peertube-plugin-livechat-button-close').style.display='none'

    document.querySelector('#video-wrapper').style.justifyContent = 'center'
    document.querySelector('my-plugin-placeholder').style.flexGrow='0'
    document.querySelector('#videojs-wrapper').style.flexGrow='1'
  }
})


document.querySelector('#video-wrapper').style.justifyContent = 'revert'
document.querySelector('my-plugin-placeholder').style.flexGrow='1'
document.querySelector('#videojs-wrapper').style.flexGrow='0'
}, 5000);
```

**Comment by Horhik** [comment]
> Can you recommend me some standard fonts?

Just use `sans-serif` on the first place for messages and titles, pls

And to hide this black gap above chat, you can set 
```  css
height: 0px;
z-index: 10;
```
for `.peertube-plugin-livechat-buttons`

BTW, are you planning to make converse.js styling more available and configurable?

**Comment by JohnXLivingston** [comment]
Thanks @Horhik , your «shitty hotfix» give me some leads for a better UI. I'll look into this further next week.



> BTW, are you planning to make converse.js styling more available and configurable?

I plan to write some custom ConverseJS templates, and rewrite the CSS.
I'm willing to use some Peertube's colors. I have to find a way to reuse colors from the current theme. I don't know yet how to achieve this. I'll try this next week.

When you say «more available and configurable», you mean for the plugin developers, or for the instance admins (to override styles)?

**Comment by Horhik** [comment]
> When you say «more available and configurable», you mean for the plugin developers, or for the instance admins (to override styles)?

I mean "for the instance admins"

Glad to hear that I could help you with UI :)

**Comment by JohnXLivingston** [comment]
> I mean "for the instance admins"

I have to think about it. Anyway, it will be the last thing to do. Otherwise, it can make new developments harder (to not break backward compatibility once it is available).



> Glad to hear that I could help you with UI :)

Yes, i'm not a CSS expert :)
I searched for some UI/ConverseJS expert to contribute several month ago, but I did not find one. So I'll have to do this myself.

**Comment by drequivalent** [comment]
7. Nice one but please make it customizable, so the instances can have their own logo and style appiled.

**Comment by JohnXLivingston** [comment]
@drequivalent , i'll do my best to take your remarks into account.

**Comment by JohnXLivingston** [comment]
Here are some «work in progress» for the new styling:
https://videos.john-livingston.fr/w/rQqh7UJRXDst6bAkJKd6Yy
And for the fullscreen version:
https://videos.john-livingston.fr/plugins/livechat/router/webchat/room/d1372116-fd32-48f7-b83e-c31029465a14

This is highly work in progress, and will be updated frequently. 
For now, there are:

- using default Peertube theme's colors. Next, i'll try to auto ajust to the user's current theme colors.
- hide avatars when width < 576px
- larger occupants' sidebar when width < 576px
- new splashscreen
- fonts = "Source Sans Pro", sans-serif. Is this ok for russian?
- more to come...

Feel free to comment.

**Comment by JohnXLivingston** [comment]
> So, I made one shitty hotfix. It is used as a custom code to add to your peertube instance. Maybe this will help you somehow.

@Horhik , I tried to include your ideas in this commit: https://github.com/JohnXLivingston/peertube-plugin-livechat/commit/5f3e8f08a2885285fed4eea91c5a764ae5352355
It seams that your `justifyContent = 'reverse'` is not a valid value. But it is not necessary.
Also, changing the flexGrow from the videojs-wrapper is also not necessary ('0' is the default value for this attribute).
I just changed the plugin placeholder flexGrow when the chat is open.

You can see the demo link in my previous content, it was updated.

I have a bug for some window dimensions. The chat goes beyond the right. See screenshot:
![image](https://user-images.githubusercontent.com/38844060/142213901-add92cad-7505-48f3-9deb-c439899d1ab4.png)
Not sure it is related to this commit. But if someone has a clue...

**Comment by JohnXLivingston** [comment]
I'm trying a new feature: close button, and «open in new window» button are now smaller when the chat is open.
Tell me what you think. Is this too small? Is this better?
![image](https://user-images.githubusercontent.com/38844060/142436173-dcb0e230-ca4f-4a54-b2af-c079aee11417.png)
![image](https://user-images.githubusercontent.com/38844060/142436221-03877f3c-c608-4163-bac9-c5bb20f5162d.png)



PS: I don't want to use the z-index trick (see comment https://github.com/JohnXLivingston/peertube-plugin-livechat/issues/72#issuecomment-962179640 ), because it can hide some content (for people using external webchat tools).

**Comment by Horhik** [comment]
> Also, changing the flexGrow from the videojs-wrapper is also not necessary
Are you sure? As I can see on your test site, the default value of `flex-grow` is `1`, but, anyway, if I'm closing the window the buttons are on the left with `flex-grow` `0` that's less intuitive than buttons which constantly sticking to the right top corner that is possible to do with `flex-grow` `1` when chat is closed
![Peek 2021-11-19 00-43](https://user-images.githubusercontent.com/46262811/142502124-ed354c6c-ca26-419e-9fcc-e86b7f553389.gif)

**Comment by JohnXLivingston** [comment]
>  Are you sure? As I can see on your test site, the default value of `flex-grow` is `1`,
Hum, you are right. I must have made a mistake. I missed the CSS that set the flex-grow to "1", and I thought it was using the default CSS flew-grow value ("0").

> but, anyway, if I'm closing the window the buttons are on the left with `flex-grow` `0` that's less intuitive than buttons which constantly sticking to the right top corner that is possible to do with `flex-grow` `1` when chat is closed

Personally, I prefer it the way it is now. If other people have an opinion, I'd love to hear it.

**Comment by JohnXLivingston** [comment]
New feature: auto color detection, depending on the peertube's current theme.
Available on https://videos.john-livingston.fr/w/rQqh7UJRXDst6bAkJKd6Yy for testing. 
NB: the page needs to be refreshed after changing the theme (or close/open the chat).

![image](https://user-images.githubusercontent.com/38844060/142867513-557041c4-c736-441b-bf40-4795834f0eb6.png)

![image](https://user-images.githubusercontent.com/38844060/142867847-4e9bbb63-a5ea-4e21-a9d2-20e8f0dffd0e.png)

![image](https://user-images.githubusercontent.com/38844060/142867928-47a6a8c7-0561-4a7d-9388-0972f365ac8f.png)

![image](https://user-images.githubusercontent.com/38844060/142868112-26f80116-c58b-40a5-a5be-5c8a2e77fbed.png)

![image](https://user-images.githubusercontent.com/38844060/142868163-0e0f293f-f368-459f-a4aa-65932e575267.png)

**Comment by JohnXLivingston** [comment]
I'm almost done with the new styling. 
The beta version is available here (use private navigation to avoid cache issue): https://videos.john-livingston.fr/w/rQqh7UJRXDst6bAkJKd6Yy
Here are some links to the fullpage result, with several peertube themes (I picked some at random):

default theme : https://videos.john-livingston.fr/plugins/livechat/5.0.0/router/webchat/room/d1372116-fd32-48f7-b83e-c31029465a14?_ac_mainForeground=%23000&_ac_mainBackground=%23fff&_ac_greyForeground=%23585858&_ac_greyBackground=%23e5e5e5&_ac_menuForeground=%23fff&_ac_menuBackground=%23000&_ac_inputForeground=%23000&_ac_inputBackground=%23fff&_ac_buttonForeground=rgb%28255%2C+255%2C+255%29&_ac_buttonBackground=%23f2690d&_ac_link=%23000&_ac_linkHover=%23000
dark : https://videos.john-livingston.fr/plugins/livechat/5.0.0/router/webchat/room/d1372116-fd32-48f7-b83e-c31029465a14?_ac_mainForeground=%23fff&_ac_mainBackground=%23111111&_ac_greyForeground=gray&_ac_greyBackground=%23e5e5e5&_ac_menuForeground=%23fff&_ac_menuBackground=%23000&_ac_inputForeground=%23fff&_ac_inputBackground=gray&_ac_buttonForeground=rgb%28255%2C+255%2C+255%29&_ac_buttonBackground=%23f2690d&_ac_link=%23fff&_ac_linkHover=%23fff
framasoft : https://videos.john-livingston.fr/plugins/livechat/5.0.0/router/webchat/room/d1372116-fd32-48f7-b83e-c31029465a14?_ac_mainForeground=%231a202c&_ac_mainBackground=%23fff&_ac_greyForeground=%23677890&_ac_greyBackground=%23e2e8f0&_ac_menuForeground=%23fff&_ac_menuBackground=%231a202c&_ac_inputForeground=%231a202c&_ac_inputBackground=%23fff&_ac_buttonForeground=rgb%28255%2C+255%2C+255%29&_ac_buttonBackground=%23c42719&_ac_link=%231a202c&_ac_linkHover=%231a202c
black-red : https://videos.john-livingston.fr/plugins/livechat/5.0.0/router/webchat/room/d1372116-fd32-48f7-b83e-c31029465a14?_ac_mainForeground=%23000000&_ac_mainBackground=%23ffffff&_ac_greyForeground=%23585858&_ac_greyBackground=%23e5e5e5&_ac_menuForeground=%23ffffff&_ac_menuBackground=%23000000&_ac_inputForeground=%23000&_ac_inputBackground=%23fff&_ac_buttonForeground=rgb%28255%2C+255%2C+255%29&_ac_buttonBackground=%23ff1e00&_ac_link=%23000000&_ac_linkHover=%23000000
Pleine page thème pacerip-light-dark-flip : https://videos.john-livingston.fr/plugins/livechat/5.0.0/router/webchat/room/d1372116-fd32-48f7-b83e-c31029465a14?_ac_mainForeground=%23a9b7d3&_ac_mainBackground=%23141515&_ac_greyForeground=%23a9b7d3&_ac_greyBackground=%23e5e5e5&_ac_menuForeground=%23a9b7d3&_ac_menuBackground=%23141515&_ac_inputForeground=%23a9b7d3&_ac_inputBackground=transparent&_ac_buttonForeground=rgb%28255%2C+255%2C+255%29&_ac_buttonBackground=%23a9b7d3&_ac_link=%23a9b7d3&_ac_linkHover=%23a9b7d3

**Comment by JohnXLivingston** [comment]
I'm closing this issue. Almost all points are fixed in v5.0.0, that will be released very soon (today or tomorrow).
If you have any comments on the v5.0.0, please open a new issue.
